Added log in service

danpottshimself opened the pull request:
modules are for future services and controllers.
Log in service added.

dpotts7 committed:
Added log in service

keithbarrow commented on scripts/proxy/controller/controller.js:
```diff
@@ -0,0 +1,10 @@
+            .controller('ApiController', ['$scope', 'AuthenticateUser', 'GameApi', 'UserLogIn',
+                    function ($scope, authenticateUser, gameApi, userLogIn) {
+                            $scope.authenticateUser = authenticateUser;
+                            $scope.gameApi = gameApi;
```
You shouldn't be exposing the game API out  - there should (at least) be a model between the API and the View
This is assuming the API is correctly named

keithbarrow commented on scripts/proxy/services/authenticate-user.js:
```diff
@@ -0,0 +1,36 @@
+(function () {
+    'use strict';
+    angular.module('Tombola.Module.ApiCall')
+        .service('AuthenticateUser', [ '$http', '$q', 'DataHandle',
```
Naming - providers are the Angular equivalent of classes, which means the names should be class like.
If you are naming using verbs (such as _Authenticate_User) this implies a function/method. Classes should be noun-named: `UserAuthentication`

keithbarrow commented on scripts/proxy/services/authenticate-user.js:
```diff
@@ -0,0 +1,36 @@
+                    return dataHandle.dataHandler('/users/logout', {}, headers, 'POST');
+                };
+
+                me.nextGameInformation = function (token) {
```
What is next game functionality doing in a service related to authentication.
What I would do: 
1. Have a Game API proxy (or, if that isn't clean enough an AuthProxy for Auth calls, as these tend to be different and one for Game calls) 
2. Have Models to model auth and game-play (including next game). These communicate with the proxy.

keithbarrow commented on scripts/proxy/services/authenticate-user.js:
```diff
@@ -0,0 +1,36 @@
+                        'userId': username,
+                        'balance':balance
+                        },
+                        headers = {
```
Headers are more or less standard across most of the calls - the repeated code is unecessary.

keithbarrow commented on scripts/proxy/services/data-handler.js:
```diff
@@ -0,0 +1,23 @@
+angular.module('Tombola.Module.ApiCall')
+    .service('DataHandle',['$http', '$q',
```
Naming: this does not handle data - so don't name it that.
It is a GameServerApiProxy
The two other proxies I mentioned above should use this - you can add header & data stuff in here dependent on whether you actually supply them in the call.

keithbarrow commented on scripts/proxy/services/log-in.js:
```diff
@@ -0,0 +1,37 @@
+    angular.module('Tombola.Module.ApiCall')
+        .service('UserLogIn', ['$state', 'AuthenticateUser', 'DataHandle',
+            function ($state, authenticateUser, dataHandle) {
+                var me  = this;
```
Again - layering: The idea is to isolate the model (which this is) as far as possible from the API. 

I'd expect the something _like_:

```
me.logIn = function(username, password){
    loginApi.logIn(username, password)
}
```

There might be some `promise` work in there - You'd probably want to keep the token in this service and a method to check whether the user is logged in.

dpotts7 committed:
Made changes as requested.
Renamed everything to more appropriate and logical names.

keithbarrow commented on scripts/proxy/services/game-timer.js:
```diff
@@ -0,0 +1,30 @@
+(function () {
+    'use strict';
+    angular.module('Tombola.Module.ApiCall')
```
I need to see you about this service, going to speak to Ian first.

keithbarrow commented on scripts/proxy/services/lobby/bingo-call-api-proxy.js:
```diff
@@ -0,0 +1,20 @@
+                        "balance": balance,
+                        "callnumber": callNumber
+                    };
+                    return logInServerApiProxy.dataHandler('/game/getcall', data, token, 'POST');
```
I thought I'd left a comment to rename data handler? It is really calling the API so `.call(...)` seems more appropriate.

keithbarrow commented on scripts/proxy/services/lobby/bingo-call-model.js:
```diff
@@ -0,0 +1,41 @@
+    'use strict';
+    angular.module('Tombola.Module.ApiCall')
+        .service('BingoCall', ['$timeout', 'LogInServerApiProxy', 'UserLogIn','CheckWinners','BingoCallApiProxy', 'TicketCreation',
+            function ($timeout, logInServerApiProxy, userLogIn, checkForWinners, bingoCallApiProxy, ticketCreation) {
```
Id loginServerApiProxy used? - If so it feels like a code smell:  Not sure why something concerned with the bingo game would be concerned with communicating with the login functionality of the game. 
Similar comment for the userLogin service: once you are logged in it'd be better to have a user service of some sort - in a proper system there would be a bunch of stuff tied in to the user - not just the balance but things like real name, e-mail avatar etc (log into the out main site and navigate to myAccount --> Update personal details). This also prevents security information bleeding through

dpotts7 committed:
renamed the data handler method

dpotts7 committed:
The user log in proxy was not being used. Silly mistake

keithbarrow commented on scripts/proxy/services/lobby/bingo-call-model.js:
```diff
@@ -0,0 +1,41 @@
+                    ticketCreation.ifNumbersMatch(me.call);
+                    calledBingoBalls();
+                    checkForWinners.checkForWinner(response);
+                    longPolling();
```
You are not long polling - so don't call it this. We're normal polling - just the messages are written in a format you would use to support long-polling.

dpotts7 committed:
changed the name of long polling to Api polling

keithbarrow commented on scripts/proxy/services/lobby/bingo-call-model.js:
```diff
@@ -0,0 +1,41 @@
+                me.bingoCall = function () {
+                    callNumber +=1;
+                    bingoCallApiProxy.bingoCallInformation(userLogIn.username, userLogIn.balance, callNumber, userLogIn.token)
+                        .then(function (response){
```
The proxy call should respond with an object you can just use, not the response. The idea is to isolate your model from the API response - if I suddenly change my API's responses, you'll potentially need to change every layer up to this point: not robust. The answer is provider  to convert the response into an object which you can just use and return that from the proxy.

keithbarrow commented on scripts/proxy/services/lobby/bingo-call-model.js:
```diff
@@ -0,0 +1,41 @@
+                        });
+                };
+
+                var calledBingoBalls = function (){
```
HAHAHA Balls.

keithbarrow commented on scripts/proxy/services/lobby/bingo-call-model.js:
```diff
@@ -0,0 +1,41 @@
+                };
+
+                var longPolling = function (){
+                    if(checkForWinners.houseWinner){
```
Check for winners becomes redundant if you get your converted object right.

keithbarrow commented on scripts/proxy/services/lobby/game-api-model.js:
```diff
@@ -0,0 +1,37 @@
+
+                me.getNextGame = function () {
+                    var promise = gameApiProxy.nextGameInformation(userLogIn.token);
+                    me.handlePromise(promise);
```
I don't think handle promise is buying you anything in this function- you are just making `handlePromise` check for a "NextGame" response.

keithbarrow commented on scripts/proxy/services/log-in/log-in-server-api-proxy.js:
```diff
@@ -0,0 +1,37 @@
+                return deferred.promise;
+            };
+
+            me.logInInformation = function (username, password) {
```
rename this to `login` - the function name should be what it does (i.e. a verb) rather than what it returns.

keithbarrow commented on scripts/proxy/services/log-in/log-in-server-api-proxy.js:
```diff
@@ -0,0 +1,37 @@
+                return me.dataHandler('/users/login', data, {}, 'POST');
+            };
+
+            me.logOutInformation = function (token) {
```
As above - rename to logout

keithbarrow commented on scripts/ticket/filters/blank-spaces.js:
```diff
@@ -0,0 +1,20 @@
+                  if (ticketNumber.number === '00') {
+                      return 'blankSpace';
+                  }
+                    console.log(ticketNumber);
```
Nope.

keithbarrow commented on scripts/ticket/filters/overlay.js:
```diff
@@ -0,0 +1,19 @@
+            function () {
+
+                return function(lineMessage, houseMessage){
+                    if (lineMessage === 'Well Done! You have won the line prize of £1') {
```
You've hardcoded the prize value. The prize for a line currently stands a 1p BTW.

keithbarrow commented on scripts/ticket/filters/overlay.js:
```diff
@@ -0,0 +1,19 @@
+                    if (lineMessage === 'Well Done! You have won the line prize of £1') {
+                        return 'modal';
+                    }
+                    if (houseMessage === 'Well Done! You have won the line prize of £5') {
```
You've hardcoded the prize value.

keithbarrow commented on scripts/ticket/services/check-wins.js:
```diff
@@ -0,0 +1,34 @@
+                    };
+
+                    me.lineWinner= function(response){
+                        me.lineMessage = 'Well Done! You have won the line prize of £' + response.payload.winnerInfo.lineprize;
```
This and the filter?

You can inject filters into modules - but I'd rather you didn't do that in this case - you are bleeding UI into the model

keithbarrow commented on scripts/ticket/services/check-wins.js:
```diff
@@ -0,0 +1,34 @@
+(function () {
+        'use strict';
+        angular.module('Tombola.Module.ApiCall')
+            .service('CheckWinners', ['$state', '$timeout', 'UserLogIn',
```
With better modelling this becomes redundant - the response object from the proxy should show whether this is a line or house winner, and by how much.

keithbarrow commented on scripts/ticket/services/ticket-creation.js:
```diff
@@ -0,0 +1,62 @@
+                    this.matched = false;
+                };
+                me.ticket = {
+                    numbers:[]
```
Don't call this numbers - which numbers? - called numbers? Card numbers?

keithbarrow commented on scripts/ticket/services/ticket-creation.js:
```diff
@@ -0,0 +1,62 @@
+                me.ticket = {
+                    numbers:[]
+                };
+                me.lines = [];
```
Personally I'd  structure a ticket like this:

```
ticket:[
    //Each inner array is a line
    [null, null, null, ......] ,//Each null is a square
    [null, null, null, ......] ,
    [null, null, null, ......] ,
    [null, null, null, ......] 
]
```

BUT I'd generate the empty ticket via looping - easier to maintain as you can see how many squares per line and how many line.

keithbarrow commented on scripts/ticket/services/ticket-creation.js:
```diff
@@ -0,0 +1,62 @@
+                me.sortTicket = function (cardNumber){
+                    var k;
+                    for(k = 0; k < cardNumber.length; k+=2) {
+                        me.ticket.numbers.push(new BingoNumber(parseInt(cardNumber[k]+cardNumber[k+1])));
```
I think you only use me.ticket.numbers in this loop - declare it in this function as `ticketNumbers` or something.
You shouldn't need to keep the string I give you with all the ticket numbers. - This model should be the canonical version of the ticket.

keithbarrow commented on scripts/ticket/services/ticket-creation.js:
```diff
@@ -0,0 +1,62 @@
+                                return me.lines[lineNumber][i];
+                            }
+                        }
+                    return {number: '00'};
```
Just return null.
Null means empty or unknown.

keithbarrow commented on scripts/ticket/services/ticket-creation.js:
```diff
@@ -0,0 +1,62 @@
+(function () {
```
A simpler (? not sure - haven't done this myself) option when it comes to dealing with a ticket:
1. Define a class to model a square - is has a value (null if nothing) and a "remaining" bool (yet to be called)
2. Define a class to model a line - an arrary on 9 squares plus a "togo value" (the number of squares in the line still to be called, and finally a line complete which is a bool (true when togo value =0)
3. A ticket class - contains the lines and keeps track of number of lines needed.

dpotts7 committed:
refactor pt1

dpotts7 committed:
Added authentication and a token (get/setter) service
Added a converter service which takes the Api response and returns an object instead.

dpotts7 committed:
edited template cache due to changes

dpotts7 committed:
Wrapped the state changing into separate methods

dpotts7 committed:
refactored code
